src/libraries/System.Private.CoreLib/src/System/Numerics/BitOperations.cs

```diff
@@ -708,5 +708,166 @@ public static nuint RotateRight(nuint value, int offset)
+
+            // Implicit byte[] to ReadOnlySpan<byte> conversion
+            ReadOnlySpan<byte> bytes = BitConverter.GetBytes(data);
+
+            foreach (byte b in bytes)
+            {
+                crc = ((crc >> 8) | (crc << 8)) & 0xFFFF;
+                crc ^= b & 0xFFu; // Implicit byte to uint conversion + Sign Truncation
+                crc ^= (crc & 0xFF) >> 4;
+                crc ^= (crc << 12) & 0xffff;
+                crc ^= ((crc & 0xFF) << 5) & 0xffff;
+            }
+
+            return crc;
+        }
+
+        /// <summary>
+        /// Calculates the CRC (Cyclic redundancy check) checksum.
+        ///
+        /// Uses following Hardware Intrinsics, if supported:
+        /// * unsigned int _mm_crc32_u32 (unsigned int crc, unsigned int v)
+        /// * uint32_t __crc32w (uint32_t a, uint32_t b)
+        /// </summary>
+        /// <param name="crc">The base value to calculate checksum on</param>
+        /// <param name="data">The checksum data</param>
+        /// <returns>The CRC-checksum</returns>
+        [CLSCompliant(false)]
+        public static uint Crc32(uint crc, uint data)
+        {
+            if (Sse42.IsSupported)
+            {
+                // unsigned int _mm_crc32_u32 (unsigned int crc, unsigned int v)
+                return Sse42.Crc32(crc, data);
+            }
+            if (Runtime.Intrinsics.Arm.Crc32.IsSupported)
+            {
+                // uint32_t __crc32w (uint32_t a, uint32_t b)
+                return Runtime.Intrinsics.Arm.Crc32.ComputeCrc32(crc, data);
+            }
+
+
+            // Software fallback
+
+            // Implicit byte[] to ReadOnlySpan<byte> conversion
+            ReadOnlySpan<byte> bytes = BitConverter.GetBytes(data);
+
+            foreach (byte b in bytes)
+            {
+                crc = ((crc >> 8) | (crc << 8)) & 0xFFFF;
+                crc ^= b & 0xFFu; // Implicit byte to uint conversion + Sign Truncation
+                crc ^= (crc & 0xFF) >> 4;
+                crc ^= (crc << 12) & 0xffff;
+                crc ^= ((crc & 0xFF) << 5) & 0xffff;
+            }
+
+            return crc;
+        }
+
+        /// <summary>
+        /// Calculates the CRC (Cyclic redundancy check) checksum.
+        ///
+        /// Uses following Hardware Intrinsics, if supported:
+        /// * unsigned __int64 _mm_crc32_u64 (unsigned __int64 crc, unsigned __int64 v)
+        /// * uint32_t __crc32d (uint32_t a, uint64_t b)
+        /// </summary>
+        /// <param name="crc">The base value to calculate checksum on</param>
+        /// <param name="data">The checksum data</param>
+        /// <returns>The CRC-checksum</returns>
+        [CLSCompliant(false)]
+        public static uint Crc32(uint crc, ulong data)
+        {
+            if (Sse42.X64.IsSupported)
+            {
+                // unsigned __int64 _mm_crc32_u64 (unsigned __int64 crc, unsigned __int64 v)
+                return (uint)Sse42.X64.Crc32(crc, data);
```

Comment:
Some tests fail here due casting ulong to uint. Is there a Sse42 X64 equivalent for calculating a Crc checksum to uint?

Reply:
There is no Crc32C support which returns `uint32_t` with data paremter of type `__int64`.
See [Intel® Intrinsics Guide](https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#techs=MMX,SSE,SSE2,SSE3,SSSE3,SSE4_1,SSE4_2,AVX,AVX2,FMA,AVX_VNNI,AVX_512,KNC,AMX,SVML,Other&text=crc)

Reply:
I removed this.

Reply:
`_mm_crc32_u64` returns `__int64`, of which only the low 32 bits are populated.  It can be truncated safely.

Reply:
@saucecontrol Does this return the Castagnoli variant?

Reply:
Yes, it's the same instruction as emitted by `_mm_crc32_u32`, just with different encoding to indicate it operates on 64 bits instead of 32.  See https://www.felixcloutier.com/x86/crc32.

Reply:
@saucecontrol I have adjusted this. Does this correspond to your suggestion?

Reply:
You don't need `Unsafe.As` there.  If you were getting errors on a normal cast from `ulong` to `uint`, you might just need an `unchecked` context, although I don't believe that's necessary in CoreLib.

Reply:
done

Reply:
Are you sure that this conversion is a good practise :smile: 

The tests are failing because of this...

`System.ArgumentException : Object of type 'System.UInt64' cannot be converted to type 'System.UInt32'.`

https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-61558-merge-4697f643e9194882ac/System.Runtime.Extensions.Tests/1/console.81399b4e.log?sv=2019-07-07&se=2021-12-06T21%3A26%3A52Z&sr=c&sp=rl&sig=bquVfSnssQk3D6h%2FTnRoDZempuihIEdiWeyeBHkNYEs%3D

Reply:
The assert is in a test because you are passing in `ulong` to a test that is taking a `uint` parameter:
```csharp
        [Theory]
        [InlineData(0, 0, 0)]
        [InlineData(0, 120, 44758)]
        [InlineData(0, ulong.MaxValue, 42721)]
        [InlineData(123, ulong.MaxValue, 27055)]
        public static void BitOps_Crc32C_ulong(uint crc, uint data, uint expected)
```

My guess is this should be `public static void BitOps_Crc32C_ulong(uint crc, ulong data, uint expected)`

Reply:
Oh that was a mistake by me 😅 Thank you

Reply:
@tannergooding Do you know why this does not work on ARM64?